git-ext/src/reference/name.rs

```diff
@@ -256,6 +279,26 @@ impl OneLevel {
     pub fn as_str(&self) -> &str {
         self.as_ref()
     }
+
+    pub fn from_qualified(Qualified(path): Qualified) -> (Self, Option<RefLike>) {
+        if path.starts_with("refs/") {
+            let mut path = path.iter().skip(1);
+            match path.next() {
```

Comment:
This match highlights the fact that `Qualified` can still be in a weird form. It could be that it has `refs/mistress` so there's a category but no path. Any thoughts on resolving this or do you think it's worth resolving? :thinking:

Reply:
> /me jumps around, hysterically chanting "do what I mean! do what I mean!"

I think this is great as-is, because the semantics of git refs is really ... in the eye of the beholder. For example `OneLevel("HEAD")` is valid, as is `Qualified("refs/HEAD")`, but its semantics depend on how you look at it (conventionally, `HEAD` would be expected to be a symref). If we wanted to capture all permutations in the type system in a general way, we would need to be able to express something like this:

```haskell
type QualifiedAndCategorised a = Ref '["refs"] ++ Category a ++ (xs :: '(NonEmpty PathComponent));
type QualifiedSym = Ref '["refs", Uppercase];
type Category a = a | '["remotes", Remote, a];
type GitCategory = "heads" | "tags" | "notes" | "replace" | "bisect" | SomeCategory;
type RadCategory = (GitCategory ∉ "replace" | "bisect" | SomeCategory) | "rad"
```

Not sure it's possible to approximate that in the type system we have, and even in a more powerful one it wouldn't be trivial. So I think we should probably look into refining types in `librad` where we understand what forms we expect / allow.

You don't need the `if else`, though: a `Qualified` _always_ starts with "refs/".

Reply:
> hysterically chanting "do what I mean! do what I mean!"

Ahahaha fair enough. But then WDYM when you have a `Qualified(refs/HEAD)` and we call `from_qualified`? Should it be `(OneLevel(HEAD), None)`? If that's the case my function is wrong.

Really, we shouldn't be able to hit the `None` case because we should have `refs/` followed by some value. Do we prefer a `panic!` or `unreachable!` for these cases?

Reply:
The version in 4297476 looks good to me